### PR TITLE
gh-103131: Convert `sys.set_asyncgen_hooks` to AC

### PR DESCRIPTION
**Pull request by sobolevn**
Refs https://github.com/python/cpython/pull/11328

CC @taleinat @serhiy-storchaka as in the original PR

* Issue: gh-103131

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2023-03-30-15-29-55.gh-issue-103131.pFNxUN.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Convert :func:`sys.getsizeof` and :func:`sys.set_asyncgen_hooks`
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Core and Builtins/2023-03-30-15-29-55.gh-issue-103131.pFNxUN.rst]: It is an internal refactoring. Such changes usually are not reflected in the NEWS file.

**Review by serhiy-storchaka — COMMENTED** [review]
What is the help() output for these functions?

**Comment by sobolevn** [comment]
Hm, indeed `help` has gotten even worth. Before:

```python
>>> import sys
>>> help(sys.getsizeof)
Help on built-in function getsizeof in module sys:

getsizeof(...)
    getsizeof(object [, default]) -> int
    
    Return the size of object in bytes.

>>> help(sys.set_asyncgen_hooks)
Help on built-in function set_asyncgen_hooks in module sys:

set_asyncgen_hooks(...)
    set_asyncgen_hooks(* [, firstiter] [, finalizer])
    
    Set a finalizer for async generators objects.

```

After:

```python
>>> import sys
>>> help(sys.getsizeof)
Help on built-in function getsizeof in module sys:

getsizeof(...)
    Return the size of object in bytes.

>>> help(sys.set_asyncgen_hooks)
Help on built-in function set_asyncgen_hooks in module sys:

set_asyncgen_hooks(...)
    Set a finalizer for async generators objects.

>>> sys.getsizeof.__text_signature__
'($module, /, object, default=<unrepresentable>)'
>>> sys.set_asyncgen_hooks
<built-in function set_asyncgen_hooks>
>>> sys.set_asyncgen_hooks.__text_signature__
'($module, /, firstiter=<unrepresentable>,\n                   finalizer=<unrepresentable>)'
```

Looks like `inspect` cannot parse `<unrepresentable>` as a part of the signature.

```python
>>> inspect.signature(sys.getsizeof)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/sobolev/Desktop/cpython/Lib/inspect.py", line 3322, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sobolev/Desktop/cpython/Lib/inspect.py", line 3066, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sobolev/Desktop/cpython/Lib/inspect.py", line 2558, in _signature_from_callable
    return _signature_from_builtin(sigcls, obj,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sobolev/Desktop/cpython/Lib/inspect.py", line 2359, in _signature_from_builtin
    return _signature_fromstr(cls, func, s, skip_bound_arg)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sobolev/Desktop/cpython/Lib/inspect.py", line 2220, in _signature_fromstr
    raise ValueError("{!r} builtin has invalid signature".format(obj))
ValueError: <built-in function getsizeof> builtin has invalid signature
```

This part also looks strange:

```
'($module, /, firstiter=<unrepresentable>,\n                   finalizer=<unrepresentable>)'
```

Why `\n                   `?

In `typeshed` we use `...` for default values that are hard / impossible to represent.
Can we use the same technique here? So, `__text_signature__` would become `'($module, /, firstiter=..., finalizer=...)'`

**Comment by sobolevn** [comment]
Ok, I've found the temporary solution. Other similar cases just copy the signature into the docs: https://github.com/python/cpython/blob/20c0f196ff63546a6a0460a46c92de12b975bc1a/Objects/odictobject.c#L1072

**Review by AlexWaygood — DISMISSED** [review]
This now has merge conflicts

File: Python/sysmodule.c [hunk 2]
```diff
@@ -1291,19 +1291,22 @@ static PyStructSequence_Desc asyncgen_hooks_desc = {
...
 
-    if (!PyArg_ParseTupleAndKeywords(
-            args, kw, "|OO", keywords,
-            &firstiter, &finalizer)) {
-        return NULL;
-    }
+    firstiter: object = NULL
+    finalizer: object = NULL
+
+set_asyncgen_hooks(* [, firstiter] [, finalizer])
```
Comment (vstinner) [inline Python/sysmodule.c]: Maybe the doc should be modified to use this signature: https://docs.python.org/dev/library/sys.html#sys.set_asyncgen_hooks

Or this docstring should use the doc signature: `sys.set_asyncgen_hooks(firstiter, finalizer)`.

Comment (erlend-aasland) [reply Python/sysmodule.c]: The docs also read:

> Accepts two optional keyword arguments [...]

IMO, the signature in the docs should be updated to reflect the actual implementation.

Comment (sobolevn) [reply Python/sysmodule.c]: Done in https://github.com/python/cpython/pull/114385

Comment (vstinner) [reply Python/sysmodule.c]: Can you please rebase your PR on top of it? There are now merge conflicts.

**Comment by vstinner** [comment]
The help now shows the signature twice. Please remove the second signature:

```
>>> help(sys.set_asyncgen_hooks)
set_asyncgen_hooks(firstiter=<unrepresentable>,
                   finalizer=<unrepresentable>)
    set_asyncgen_hooks([firstiter] [, finalizer])

    Set a finalizer for async generators objects.

>>> help(sys.getsizeof)
getsizeof(object, default=<unrepresentable>)
    getsizeof(object [, default]) -> int

    Return the size of object in bytes.
```

File: Python/sysmodule.c [hunk 3]
```diff
@@ -1903,20 +1900,25 @@ _PySys_GetSizeOf(PyObject *o)
     return (size_t)size + presize;
 }
 
+/*[clinic input]
+sys.getsizeof
+
+    object: object
+    default as dflt: object = NULL
```
Comment (vstinner) [inline Python/sysmodule.c]: I suggest to use a longer name rather than a shorter one:

```suggestion
    default as default_size: object = NULL
```

**Comment by sobolevn** [comment]
The `sys.getsizeof` change is incorrect. I can still convert `sys.set_asyncget_hooks` and change the defaults from `NULL` to `Py_None`, looks like that this will be valid.

File: Python/sysmodule.c (hunk 3 above)
Comment (erlend-aasland) [reply Python/sysmodule.c]: That would make the diff unnecessary bigger. IMO, it is better to align with the current naming.

**Comment by sobolevn** [comment]
Here's the new `help()` output:

```python
» ./python.exe
Python 3.13.0a5+ (heads/issue-103131:c0086505210, Mar 20 2024, 13:48:58) [Clang 15.0.0 (clang-1500.0.40.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> help(sys.set_asyncgen_hooks)
Help on built-in function set_asyncgen_hooks in module sys:

set_asyncgen_hooks(firstiter=None, finalizer=None)
    Set a finalizer for async generators objects.
```

`None` is supported and even tested here: https://github.com/python/cpython/blob/8182319de33a9519a2f243ac8c35a20ef82a4d2d/Lib/test/test_sys.py#L1766-L1790

So, this now should be a simple change :)

**Comment by sobolevn** [comment]
Ok, I've found this branch:

```
else if (finalizer == Py_None && _PyEval_SetAsyncGenFinalizer(NULL) < 0) {
        return NULL;
    }
```

Looks like this PR is not correct either. It is better to close it.

**Pull request closed without merging**